Add DPCPP_ENABLE_OPAQUE_POINTERS option

bader opened the pull request:
This option controls the default behavior of LLVM projects regarding
opaque pointers. The default value is OFF as most of the SYCL compiler
components are not ready to handle opaque pointers.

This change adds a new nightly job testing SYCL compiler with enabled
opaque pointers by default.

andykaylor commented on llvm/CMakeLists.txt:
```diff
@@ -803,6 +803,12 @@ if(NOT LLVM_ENABLE_NEW_PASS_MANAGER)
                       " no longer supported.")
 endif()
 
+set(LLVM_ENABLE_OPAQUE_POINTERS FALSE CACHE BOOL
```
What would you think about using a different name to make it clear that this variable is specific to the intel/llvm fork?

bader replied:
I've renamed the option to `DPCPP_ENABLE_OPAQUE_POINTERS`.

andykaylor commented on clang/CMakeLists.txt:
```diff
@@ -252,7 +252,7 @@ set(CLANG_SPAWN_CC1 OFF CACHE BOOL
 
 # Manually handle default so we can change the meaning of a cached default.
-set(CLANG_ENABLE_OPAQUE_POINTERS "DEFAULT" CACHE STRING
+set(CLANG_ENABLE_OPAQUE_POINTERS ${LLVM_ENABLE_OPAQUE_POINTERS} CACHE STRING
```
It's not entirely clear to me what the upstream code is trying to accomplish with the cached default, but I think this change breaks it. Would modifying line 258 be better?

bader replied:
Done, but I don't think there a practical difference for the project.

andykaylor commented on clang/test/CodeGenOpenCL/sampled_image.cl:
```diff
@@ -1,5 +1,5 @@
-// RUN: %clang_cc1 %s -triple x86_64-unknown-linux-gnu -O0 -emit-llvm -o - | FileCheck %s
-// RUN: %clang_cc1 %s -triple x86_64-unknown-linux-gnu -O0 -emit-llvm -o - -cl-std=clc++ | FileCheck %s
+// RUN: %clang_cc1 %s -triple x86_64-unknown-linux-gnu -no-opaque-pointers -O0 -emit-llvm -o - | FileCheck %s
```
What's the value of running this test with '-no-opaque-pointers' in builds where opaque pointers are enabled? Why not just mark it as unsupported in that case?

bader replied:
I did this to follow the approach to handle the rest of OpenCL tests, which are not switched to opaque pointers yet.
Marking unsupported won't work because this test invokes the front-end compiler. We don't adjust front-end default for opaque pointers support (see clang/include/clang/Driver/Options.td), so we can change it only with explicit option. As you can see from the test results just a few tests requires adjustment and the rest of them are already updated to work with ON default value for opaque pointers support in FE. Unsupported mark works for the test invoking the driver.

It looked more closely at the tests and realized that CodeGenSPIRV tests are outdated, so I removed them (see commit messages for https://github.com/intel/llvm/pull/6378/commits/a926b481e458fe4dea16d943fceef74965fde63c and https://github.com/intel/llvm/pull/6378/commits/b41c3c70d0912cced066c653b726fe6626251a4d for more details). I updated the checks in the rest of the tests to handle opaque pointers (including the OpenCL test we added).

pvchupin reviewed (approved):
CI changes looks ok

premanandrao commented on clang/test/CodeGenSYCL/debug-info-srcpos-kernel.cpp:
```diff
@@ -25,7 +25,7 @@ int main() {
 
 // CHECK: define{{.*}} spir_kernel {{.*}}19use_kernel_for_test({{.*}}){{.*}} !dbg [[KERNEL:![0-9]+]] {{.*}}{
-// CHECK: getelementptr inbounds %class.anon, %class.anon addrspace(4)* %{{.*}}, i32 0, i32 0, !dbg [[LINE_A0:![0-9]+]]
+// CHECK: getelementptr inbounds %class.anon, {{.*}}, i32 0, i32 0, !dbg [[LINE_A0:![0-9]+]]
```
```suggestion
// CHECK: getelementptr inbounds %class.anon, ptr addrspace(4) %{{.*}}, i32 0, i32 0, !dbg [[LINE_A0:![0-9]+]]
```
That's how I modified other opaque pointer related tests in the past.   This keeps checking the 'addrspace(4)', if that was important to begin with.  Just a suggestion.

bader replied:
Thanks for the suggestion.
While we transition to opaque pointers this test is expected to work in both modes. `ptr` check will work only when opaque pointers support is enabled.
Also I don't think that checking second argument type is important for this test. The main goal of this test is to validate debug info metadata.

asudarsa reviewed (approved):
'Tools' related changes look ok to me. Thanks

premanandrao reviewed (approved):
FE changes look okay to me.

bader commented:
@andykaylor, @asudarsa, could you take a look at new commits, please?

bader commented:
> @andykaylor, @asudarsa, could you take a look at new commits, please?

Ping.

asudarsa commented:
Hi @bader 
Sorry for delay in response. LGTM except for one issue I have highlighted.
Thanks

asudarsa commented on llvm/CMakeLists.txt:
```diff
@@ -804,11 +804,11 @@ if(NOT LLVM_ENABLE_NEW_PASS_MANAGER)
 endif()
 
-set(LLVM_ENABLE_OPAQUE_POINTERS FALSE CACHE BOOL
+set(DPCPP_ENABLE_OPAQUE_POINTERS FALSE CACHE BOOL
```
Hmm..the change looks different (good) when seen as part of "changes from all commits".

asudarsa reviewed (dismissed):
LGTM. Thanks

bader commented:
@andykaylor, this PR holds off follow-up fixes, so I'd like to merge it as soon as possible. Is it okay to commit current version and address further comments (if any) separately?

bader commented:
@jchlanda, do you think that this failure on NVIDIA GPU https://github.com/intel/llvm/runs/7391421134?check_suite_focus=true is related to this change? This change is not expected to change the default behavior.

bader commented:
@jcranmer-intel, I've added https://github.com/intel/llvm/pull/6378/commits/ab0502149aa3eb107a1b7cce6ca496c9bc7341a1, which updates a few tests.
* sycl/test/check_device_code/id_queries_fit_int.cpp - now checks are invariant to opaque pointer setting (perfect case).
* sycl/test/check_device_code/kernel_arguments_as.cpp - the purpose of the checks is to validate that address space information is passed via LLVM IR and pointer type is not relevant for this test. I forced opaque pointers use and updated the checks.
* sycl/test/check_device_code/no_offset.cpp - what is interesting in this test is that compiler is able to remove certain kernel parameters if `nooffset` accessor property is set. Pointer types are not relevant here as well, so I applied the same approach as in sycl/test/check_device_code/kernel_arguments_as.cpp test.
* sycl/test/extensions/sub_group_as.cpp - this is an example of the test where further test updates are required. Type are not important for some of the checks (e.g. spirv_GenericCastToPtrExplicit_ToLocal doesn't change the type, just an address space), but there are examples where LLVM IR must have type information somewhere. I'm hesitant to update more tests like this because not all changes are in place to make the final test fix. We need to figure out what is the plan for ESIMD/matrix types and decide whether we go with https://github.com/intel/llvm/pull/6397 before updating LLVM IR check. Let me know your thoughts on this.

jchlanda commented:
> @jchlanda, do you think that this failure on NVIDIA GPU https://github.com/intel/llvm/runs/7391421134?check_suite_focus=true is related to this change? This change is not expected to change the default behavior.

It doesn't look related, even if we ignore the assert messages which are a bit flaky, upon hitting the assert PI runtime should generate error `710`:

```
PI CUDA ERROR:
        Value:           710
        Name:            CUDA_ERROR_ASSERT
```

but this one reports `-999` - `Unknown PI error` (https://github.com/intel/llvm/runs/7391421134?check_suite_focus=true#step:8:1168)

bader commented:
> > @jchlanda, do you think that this failure on NVIDIA GPU https://github.com/intel/llvm/runs/7391421134?check_suite_focus=true is related to this change? This change is not expected to change the default behavior.
> 
> It doesn't look related, even if we ignore the assert messages which are a bit flaky, upon hitting the assert PI runtime should generate error `710`:
> 
> ```
> PI CUDA ERROR:
>         Value:           710
>         Name:            CUDA_ERROR_ASSERT
> ```
> 
> but this one reports `-999` - `Unknown PI error` (https://github.com/intel/llvm/runs/7391421134?check_suite_focus=true#step:8:1168)

I can't reproduce this issue locally, so I restarted CI jobs and it failed again, but with the different error:

> error: command failed with exit status: 255

https://github.com/intel/llvm/runs/7450227286?check_suite_focus=true

These look like flaky llvm-test-suite issues not caused by this patch. I'll file a report to investigate them and merge this PR.
